lib/spack/spack/spec.py

```diff
@@ -2291,12 +2291,17 @@ def from_dict(data):
         """
         # Legacy specfile format
         if isinstance(data["spec"], list):
-            return SpecfileV1.load(data)
+            spec = SpecfileV1.load(data)
+        elif int(data["spec"]["_meta"]["version"]) == 2:
+            spec = SpecfileV2.load(data)
+        else:
+            spec = SpecfileV3.load(data)
 
-        specfile_version = int(data["spec"]["_meta"]["version"])
-        if specfile_version == 2:
-            return SpecfileV2.load(data)
-        return SpecfileV3.load(data)
+        # Any git version should
+        for s in spec.traverse():
+            s.attach_git_version_lookup()
```

Comment:
Should _validate_version from the database really live here, right after we read specs in from some serialized format?  I think this covers lock files, build caches, upstreams, etc.  I also think it avoids in-memory parses, CLI stuff, and spec literals in YAML configs.